cranelift/codegen/src/isa/aarch64/lower.rs

```diff
@@ -142,29 +142,20 @@ pub(crate) fn input_to_shiftimm<C: LowerCtx<I = Inst>>(
     input_to_const(ctx, input).and_then(ShiftOpShiftImm::maybe_from_shift)
 }
 
-pub(crate) fn output_to_const_f128<C: LowerCtx<I = Inst>>(
-    ctx: &mut C,
-    out: InsnOutput,
-) -> Option<u128> {
-    if out.output > 0 {
-        None
-    } else {
-        let inst_data = ctx.data(out.insn);
-
-        match inst_data {
-            &InstructionData::UnaryConst {
-                opcode: _,
-                constant_handle,
-            } => {
-                let mut bytes = [0u8; 16];
-                let c = ctx.get_constant_data(constant_handle).clone().into_vec();
-                assert_eq!(c.len(), 16);
-                bytes.copy_from_slice(&c);
-                Some(u128::from_le_bytes(bytes))
-            }
-            _ => None,
-        }
-    }
+pub(crate) fn const_param_to_u128<C: LowerCtx<I = Inst>>(ctx: &mut C, inst: IRInst) -> u128 {
+    let data = match ctx.data(inst) {
+        &InstructionData::Shuffle { mask, .. } => ctx.get_immediate(mask),
+        &InstructionData::UnaryConst {
+            constant_handle, ..
+        } => ctx.get_constant_data(constant_handle),
+        _ => unreachable!(),
```

Comment:
I'd prefer for this helper to retain the `Option<u128>` return value -- it forces the caller to handle the wrong-instruction-type case explicitly with something like `const_param_to_u128(...).expect("Invalid const insn for i128 const")` if it should be infallible, or do something else if the input may or may not be a constant.

Actually, ideally, we'd handle `u128` constants with the same machinery as `u64` ones, via the `constant` field on `LowerCtx::get_input()`'s returned information. This would let us dedup and handle constant pools in the machine-independent code. I'm happy to defer that to later, though!

Reply:
Just to check my understanding of your idea (otherwise I'd go for the simpler solution for the time being) - I don't think that it would help us for `Opcode::Shuffle` because the mask is not an input from the point of view of `LowerCtx::get_input()` (`Opcode::Vconst` is different, of course).

As for the constant pool improvements - yes, I was keeping that in mind and I think that concentrating the constant handling in as few places as possible would help eventually, hence why I made the changes to this helper. In fact I don't think I understand the practical distinction between `DataFlowGraph::constants` and `DataFlowGraph::immediates` - at least in the cases of `Opcode::Vconst` and `Opcode::Shuffle` the handling is pretty much equivalent.

Reply:
Yeah, I am fine with the design suggested in the first paragraph -- just return `Option<u128>`. I need to do more work to make use of the constant-pool functionality in `MachBuffer`, so we can revisit this later.